crates/bitwarden-crypto/src/chacha20.rs

```diff
@@ -0,0 +1,218 @@
+use chacha20::{
+    cipher::{NewCipher, StreamCipher},
+    XChaCha20,
+};
+use chacha20poly1305::{AeadCore, AeadInPlace, KeyInit, XChaCha20Poly1305};
+use generic_array::GenericArray;
+use poly1305::{universal_hash::UniversalHash, Poly1305};
+use subtle::ConstantTimeEq;
+
+/**
+ * Note:
+ * XChaCha20Poly1305 encrypts data, and authenticates associated data using
+ * XChaCha20Poly1305 Specifically, this uses the CTX construction, proposed here: https://par.nsf.gov/servlets/purl/10391723
+ * using blake3 as the cryptographic hash function. The entire construction is called
+ * XChaCha20Poly1305Blake3CTX. This provides not only key-commitment, but full-commitment.
+ * In total, this scheme prevents attacks such as invisible salamanders.
+ */
+use crate::CryptoError;
+
+pub struct XChaCha20Poly1305Blake3CTXCiphertext {
+    nonce: [u8; 24],
+    tag: [u8; 32],
+    encrypted_data: Vec<u8>,
+    authenticated_data: Vec<u8>,
+}
+
+pub fn encrypt_xchacha20_poly1305_blake3_ctx(
+    key: &[u8; 32],
+    plaintext_secret_data: &[u8],
+    authenticated_data: &[u8],
+) -> Result<XChaCha20Poly1305Blake3CTXCiphertext, CryptoError> {
+    encrypt_xchacha20_poly1305_blake3_ctx_internal(
+        rand::thread_rng(),
+        key,
+        plaintext_secret_data,
+        authenticated_data,
+    )
+}
+
+fn encrypt_xchacha20_poly1305_blake3_ctx_internal(
+    rng: impl rand::CryptoRng + rand::RngCore,
+    key: &[u8; 32],
+    plaintext_secret_data: &[u8],
+    associated_data: &[u8],
+) -> Result<XChaCha20Poly1305Blake3CTXCiphertext, CryptoError> {
+    let nonce = XChaCha20Poly1305::generate_nonce(rng);
+    // This should never fail because XChaCha20Poly1305::generate_nonce always returns 24 bytes
+    let nonce_slice: [u8; 24] = nonce.as_slice().try_into().expect("Nonce is 24 bytes");
+
+    // This buffer contains the plaintext, that will be encrypted in-place
+    let mut buffer = Vec::from(plaintext_secret_data);
+    let cipher = XChaCha20Poly1305::new(GenericArray::from_slice(key));
+
+    let poly1305_tag = cipher
```

Comment:
Do we want different errors here, or should all be the same? (in the latter case, it might need to be a new generic "symmetricdecryptionfailed")